File: applications/test_scripts/Real-Time Shipment Status Updates.js
describe('Real-Time Shipment Status Updates', () => {
  const shipmentID = '12345';
  const userID = '67890';

  before(() => {
    cy.login(userID);
  });

  it('should navigate to the shipment tracking page', () => {
    cy.visit('/shipment-tracking');
    cy.url().should('include', '/shipment-tracking');
  });

  it('should enter the Shipment ID and display shipment details', () => {
    cy.get(ShipmentTrackingPage.trackingField).type(shipmentID);
    cy.get(ShipmentTrackingPage.submitButton).click();
    cy.get(ShipmentTrackingPage.shipmentDetails).should('contain', shipmentID);
  });

  it('should check the current status of the shipment', () => {
    cy.get(ShipmentTrackingPage.currentStatus).should('contain', 'In Transit');
  });

  it('should simulate status update to Out for Delivery', () => {
    cy.simulateStatusUpdate(shipmentID, 'Out for Delivery');
    cy.get(ShipmentTrackingPage.currentStatus).should('contain', 'Out for Delivery');
  });

  it('should simulate status update to Delivered', () => {
    cy.simulateStatusUpdate(shipmentID, 'Delivered');
    cy.get(ShipmentTrackingPage.currentStatus).should('contain', 'Delivered');
  });

  it('should verify the timestamp of the latest status update', () => {
    cy.get(ShipmentTrackingPage.timestamp).should('contain', new Date().toLocaleTimeString());
  });

  it('should refresh the page and verify status remains Delivered', () => {
    cy.reload();
    cy.get(ShipmentTrackingPage.currentStatus).should('contain', 'Delivered');
  });

  it('should log out and log back in, status remains Delivered', () => {
    cy.logout();
    cy.login(userID);
    cy.get(ShipmentTrackingPage.currentStatus).should('contain', 'Delivered');
  });

  it('should check notification settings for shipment updates', () => {
    cy.get(ShipmentTrackingPage.notificationSettings).should('be.checked');
  });

  it('should simulate network issue and attempt to update status', () => {
    cy.simulateNetworkIssue();
    cy.simulateStatusUpdate(shipmentID, 'Delivered');
    cy.get(ShipmentTrackingPage.currentStatus).should('contain', 'Delivered');
  });

  it('should verify the shipment history log', () => {
    cy.get(ShipmentTrackingPage.historyLog).should('contain', 'Delivered');
  });

  it('should check for any error messages during status updates', () => {
    cy.get(ShipmentTrackingPage.errorMessages).should('not.exist');
  });

  it('should attempt to update status from a different device', () => {
    cy.simulateStatusUpdateFromDifferentDevice(shipmentID, 'Delivered');
    cy.get(ShipmentTrackingPage.currentStatus).should('contain', 'Delivered');
  });

  it('should verify shipment status on a mobile device', () => {
    cy.viewport('iphone-x');
    cy.visit('/shipment-tracking');
    cy.get(ShipmentTrackingPage.currentStatus).should('contain', 'Delivered');
  });

  it('should check shipment status after a system reboot', () => {
    cy.systemReboot();
    cy.visit('/shipment-tracking');
    cy.get(ShipmentTrackingPage.currentStatus).should('contain', 'Delivered');
  });
});

File: bdd/step_definitions/Real-Time Shipment Status Updates.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import com.page_objects.ShipmentTrackingPage;
import java.util.*;

public class ShipmentTrackingSteps extends ShipmentTrackingPage {

    private TestHarness testHarness = new TestHarness();

    @Given("the user is logged into the system")
    public void theUserIsLoggedIntoTheSystem() {
        loginToSystem();
    }

    @When("the user navigates to the shipment tracking page")
    public void theUserNavigatesToTheShipmentTrackingPage() {
        navigateToShipmentTrackingPage();
    }

    @Then("the shipment tracking page is displayed")
    public void theShipmentTrackingPageIsDisplayed() {
        verifyShipmentTrackingPageDisplayed();
    }

    @Given("shipment data is available in the system")
    public void shipmentDataIsAvailableInTheSystem() {
        verifyShipmentDataAvailability();
    }

    @When("the user enters Shipment ID \"12345\" in the tracking field")
    public void theUserEntersShipmentIDInTheTrackingField() {
        enterShipmentID("12345");
    }

    @Then("shipment details for ID \"12345\" are displayed")
    public void shipmentDetailsForIDAreDisplayed() {
        verifyShipmentDetailsDisplayed("12345");
    }

    @Given("the user has entered the Shipment ID \"12345\"")
    public void theUserHasEnteredTheShipmentID() {
        enterShipmentID("12345");
    }

    @When("the user checks the current status of the shipment")
    public void theUserChecksTheCurrentStatusOfTheShipment() {
        checkCurrentShipmentStatus();
    }

    @Then("the current status is displayed as 'In Transit'")
    public void theCurrentStatusIsDisplayedAsInTransit() {
        verifyCurrentStatus("In Transit");
    }

    @Given("the current status is 'In Transit'")
    public void theCurrentStatusIsInTransit() {
        verifyCurrentStatus("In Transit");
    }

    @When("a status update is simulated to 'Out for Delivery'")
    public void aStatusUpdateIsSimulatedToOutForDelivery() {
        simulateStatusUpdate("Out for Delivery");
    }

    @Then("the status updates to 'Out for Delivery' in real-time")
    public void theStatusUpdatesToOutForDeliveryInRealTime() {
        verifyStatusUpdate("Out for Delivery");
    }

    @Given("the current status is 'Out for Delivery'")
    public void theCurrentStatusIsOutForDelivery() {
        verifyCurrentStatus("Out for Delivery");
    }

    @When("a status update is simulated to 'Delivered'")
    public void aStatusUpdateIsSimulatedToDelivered() {
        simulateStatusUpdate("Delivered");
    }

    @Then("the status updates to 'Delivered' in real-time")
    public void theStatusUpdatesToDeliveredInRealTime() {
        verifyStatusUpdate("Delivered");
    }

    @Given("the status is 'Delivered'")
    public void theStatusIsDelivered() {
        verifyCurrentStatus("Delivered");
    }

    @When("the user verifies the timestamp of the latest status update")
    public void theUserVerifiesTheTimestampOfTheLatestStatusUpdate() {
        verifyTimestampOfLatestStatusUpdate();
    }

    @Then("the timestamp matches the current time")
    public void theTimestampMatchesTheCurrentTime() {
        verifyTimestampMatchesCurrentTime();
    }

    @When("the user refreshes the page")
    public void theUserRefreshesThePage() {
        refreshPage();
    }

    @Then("the status remains 'Delivered' with the same timestamp")
    public void theStatusRemainsDeliveredWithTheSameTimestamp() {
        verifyStatusAndTimestamp("Delivered");
    }

    @Given("the user is logged out")
    public void theUserIsLoggedOut() {
        logout();
    }

    @When("the user logs back in")
    public void theUserLogsBackIn() {
        loginToSystem();
    }

    @Then("the shipment status remains 'Delivered'")
    public void theShipmentStatusRemainsDelivered() {
        verifyCurrentStatus("Delivered");
    }

    @Given("the user is logged in")
    public void theUserIsLoggedIn() {
        loginToSystem();
    }

    @When("the user checks the notification settings for shipment updates")
    public void theUserChecksTheNotificationSettingsForShipmentUpdates() {
        checkNotificationSettings();
    }

    @Then("notifications are enabled for real-time updates")
    public void notificationsAreEnabledForRealTimeUpdates() {
        verifyNotificationsEnabled();
    }

    @Given("the user is attempting to update status")
    public void theUserIsAttemptingToUpdateStatus() {
        attemptStatusUpdate();
    }

    @When("a network issue is simulated")
    public void aNetworkIssueIsSimulated() {
        simulateNetworkIssue();
    }

    @Then("the system handles the issue gracefully and retries the update")
    public void theSystemHandlesTheIssueGracefullyAndRetriesTheUpdate() {
        verifyNetworkIssueHandling();
    }

    @Given("the status updates have been made")
    public void theStatusUpdatesHaveBeenMade() {
        verifyStatusUpdatesMade();
    }

    @When("the user verifies the shipment history log")
    public void theUserVerifiesTheShipmentHistoryLog() {
        verifyShipmentHistoryLog();
    }

    @Then("the log contains all status updates with timestamps")
    public void theLogContainsAllStatusUpdatesWithTimestamps() {
        verifyLogContainsAllUpdates();
    }

    @When("the user checks for any error messages")
    public void theUserChecksForAnyErrorMessages() {
        checkForErrorMessages();
    }

    @Then("no error messages are displayed")
    public void noErrorMessagesAreDisplayed() {
        verifyNoErrorMessages();
    }

    @When("the user attempts to update status from a different device")
    public void theUserAttemptsToUpdateStatusFromADifferentDevice() {
        attemptStatusUpdateFromDifferentDevice();
    }

    @Then("the status updates are synchronized across devices")
    public void theStatusUpdatesAreSynchronizedAcrossDevices() {
        verifyStatusSynchronization();
    }

    @When("the user verifies the shipment status on a mobile device")
    public void theUserVerifiesTheShipmentStatusOnAMobileDevice() {
        verifyShipmentStatusOnMobileDevice();
    }

    @Then("the shipment status is consistent with the desktop view")
    public void theShipmentStatusIsConsistentWithTheDesktopView() {
        verifyStatusConsistency();
    }

    @When("the system reboots")
    public void theSystemReboots() {
        rebootSystem();
    }

    @Then("the shipment status remains available")
    public void theShipmentStatusRemainsAvailable() {
        verifyShipmentStatusAvailability();
    }
}
